sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,505 @@
+            BatchPayloadSize = batchPayloadSize ?? SearchIndexingBufferedSenderOptions<T>.DefaultBatchPayloadSize;
+            RetryCount = retryCount ??  SearchIndexingBufferedSenderOptions<T>.DefaultRetryCount;
+
+            // Start the message loop
+            _readerLoop = Task.Run(ProcessMessagesAsync, publisherCancellationToken);
+        }
+
+        #region Message Producers - only called externally and do not touch state
+        /// <summary>
+        /// Add documents to be published.  This should only be called outside
+        /// of the Publisher itself.
+        /// </summary>
+        /// <param name="documents">The documents to publish.</param>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task AddDocumentsAsync(IEnumerable<T> documents, bool async, CancellationToken cancellationToken = default)
+        {
+            EnsureNotDisposed();
+            await _channel.Writer.WriteInternal(
+                Message.Publish(documents),
+                async,
+                cancellationToken)
+                .ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Flush any pending documents.  This should only be called outside of
+        /// the Publisher itself.
+        /// </summary>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task FlushAsync(bool async, CancellationToken cancellationToken = default)
+        {
+            EnsureNotDisposed();
+            await FlushInternal(async, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Flush any pending documents.  This will not check for disposal.
+        /// </summary>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        private async Task FlushInternal(bool async, CancellationToken cancellationToken = default)
+        {
+            Task flushCompletionTask = _flushCompletionSource.Task;
+            await _channel.Writer.WriteInternal(
+                Message.Flush(),
+                async,
+                cancellationToken)
+                .ConfigureAwait(false);
+            if (async)
+            {
+                await flushCompletionTask.ConfigureAwait(false);
+            }
+            else
+            {
+                #pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                flushCompletionTask.EnsureCompleted();
```

Comment:
Wouldn't this throw for non-completed tasks?

Reply:
Yes - and changed anyway now that I'm using `AwaitWithCancellation`.